// @volar/vue-code-gen/out/generators/template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPatchForSlotNode = exports.generate = exports.transformContext = void 0;
const SourceMaps = require("@volar/source-map");
const code_gen_1 = require("@volar/code-gen");
const shared_1 = require("@vue/shared");
const CompilerDOM = require("@vue/compiler-dom");
const CompilerCore = require("@vue/compiler-core");
const transform_1 = require("../transform");
const capabilitiesSet = {
    all: { basic: true, diagnostic: true, references: true, definitions: true, rename: true, completion: true, semanticTokens: true },
    noDiagnostic: { basic: true, references: true, definitions: true, rename: true, completion: true, semanticTokens: true },
    diagnosticOnly: { diagnostic: true, completion: true },
    tagHover: { basic: true },
    event: { basic: true, diagnostic: true },
    tagReference: { references: true, definitions: true, rename: { in: false, out: true } },
    attr: { basic: true, diagnostic: true, references: true, definitions: true, rename: true },
    attrReference: { references: true, definitions: true, rename: true },
    scopedClassName: { references: true, definitions: true, rename: true, completion: true },
    slotName: { basic: true, diagnostic: true, references: true, definitions: true, completion: true },
    slotNameExport: { basic: true, diagnostic: true, references: true, definitions: true, /* referencesCodeLens: true */ },
    refAttr: { references: true, definitions: true, rename: true },
};
const formatBrackets = {
    empty: ['', ''],
    round: ['(', ')'],
    // fix https://github.com/johnsoncodehk/volar/issues/1210
    curly: ['({ __VLS_foo:', '})'],
    square: ['[', ']'],
};
const validTsVar = /^[a-zA-Z_$][0-9a-zA-Z_$]*$/;
// @ts-ignore
exports.transformContext = {
    onError: () => { },
    helperString: str => str.toString(),
    replaceNode: node => { },
    cacheHandlers: false,
    prefixIdentifiers: false,
    scopes: {
        vFor: 0,
        vOnce: 0,
        vPre: 0,
        vSlot: 0,
    },
    expressionPlugins: ['typescript'],
};
function generate(ts, sourceLang, templateAst, isVue2, experimentalRuntimeMode, allowTypeNarrowingInEventExpressions, cssScopedClasses = [], htmlToTemplate, searchTexts) {
    const tsCodeGen = new code_gen_1.CodeGen();
    const tsFormatCodeGen = new code_gen_1.CodeGen();
    const cssCodeGen = new code_gen_1.CodeGen();
    const attrNames = new Set();
    const slots = new Map();
    const slotExps = new Map();
    const cssScopedClassesSet = new Set(cssScopedClasses);
    const tags = {};
    const tagResolves = {};
    const localVars = {};
    const identifiers = new Set();
    const scopedClasses = [];
    const blockConditions = [];
    tsFormatCodeGen.addText('export { };\n');
    let elementIndex = 0;
    for (const childNode of templateAst.children) {
        collectTags(childNode);
    }
    for (const tagName in tags) {
        const tag = tags[tagName];
        const tagRanges = tag.offsets.map(offset => ({ start: offset, end: offset + tagName.length }));
        const isNamespacedTag = tagName.indexOf('.') >= 0;
        const var_correctTagName = `__VLS_${elementIndex++}`;
        const var_rawComponent = `__VLS_${elementIndex++}`;
        const var_slotsComponent = `__VLS_${elementIndex++}`;
        const var_emit = `__VLS_${elementIndex++}`;
        const var_slots = `__VLS_${elementIndex++}`;
        if (isNamespacedTag) {
            for (let i = 0; i < tagRanges.length; i++) {
                const tagRange = tagRanges[i];
                if (i === 0) {
                    tsCodeGen.addText(`declare const ${var_rawComponent}: typeof `);
                }
                else {
                    tsCodeGen.addText(`declare const __VLS_${elementIndex++}: typeof `);
                }
                writeCode(tagName, tagRange, SourceMaps.Mode.Offset, {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.all,
                });
                tsCodeGen.addText(`;\n`);
            }
        }
        else {
            tsCodeGen.addText(`declare const ${var_correctTagName}: __VLS_types.GetComponentName<typeof __VLS_rawComponents, '${tagName}'>;\n`);
            tsCodeGen.addText(`declare const ${var_rawComponent}: __VLS_types.GetProperty<typeof __VLS_rawComponents, typeof ${var_correctTagName}, any>;\n`);
        }
        tsCodeGen.addText(`declare const ${var_slotsComponent}: __VLS_types.SlotsComponent<typeof ${var_rawComponent}>;\n`);
        tsCodeGen.addText(`declare const ${var_emit}: __VLS_types.ExtractEmit2<typeof ${var_rawComponent}>;\n`);
        tsCodeGen.addText(`declare const ${var_slots}: __VLS_types.DefaultSlots<typeof ${var_rawComponent}>;\n`);
        const name1 = tagName; // hello-world
        const name2 = (0, shared_1.camelize)(tagName); // helloWorld
        const name3 = name2[0].toUpperCase() + name2.slice(1); // HelloWorld
        const componentNames = new Set([name1, name2, name3]);
        if (!isNamespacedTag) {
            // split tagRanges to fix end tag definition original select range mapping to start tag
            for (const tagRange of tagRanges) {
                tsCodeGen.addText(`// @ts-ignore\n`);
                tsCodeGen.addText(`({ `);
                writeObjectProperty2(tagName, [tagRange], {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.tagHover,
                });
                tsCodeGen.addText(`: {} as `);
                tsCodeGen.addText(`__VLS_types.PickNotAny<`.repeat(componentNames.size - 1));
                const names = [...componentNames];
                for (let i = 0; i < names.length; i++) {
                    if (i > 0) {
                        tsCodeGen.addText(', ');
                    }
                    tsCodeGen.addText(`typeof __VLS_rawComponents`);
                    writePropertyAccess2(names[i], [tagRange], {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.tagReference,
                        normalizeNewName: tagName === names[i] ? undefined : unHyphenatComponentName,
                        applyNewName: keepHyphenateName,
                    });
                    if (i > 0) {
                        tsCodeGen.addText('>');
                    }
                }
                tsCodeGen.addText(` });\n`);
            }
        }
        /* Completion */
        tsCodeGen.addText('/* Completion: Emits */\n');
        for (const name of componentNames) {
            tsCodeGen.addText('// @ts-ignore\n');
            tsCodeGen.addText(`${var_emit}('${searchTexts.getEmitCompletion(name)}');\n`);
        }
        tsCodeGen.addText('/* Completion: Props */\n');
        for (const name of componentNames) {
            tsCodeGen.addText('// @ts-ignore\n');
            tsCodeGen.addText(`(<${var_rawComponent} ${searchTexts.getPropsCompletion(name)}/>);\n`);
        }
        tagResolves[tagName] = {
            rawComponent: var_rawComponent,
            slotsComponent: var_slotsComponent,
            emit: var_emit,
            slots: var_slots,
            offsets: tag.offsets.map(offset => { var _a; return (_a = htmlToTemplate(offset, offset)) === null || _a === void 0 ? void 0 : _a.start; }).filter(notEmpty),
        };
    }
    for (const childNode of templateAst.children) {
        tsCodeGen.addText(`{\n`);
        visitNode(childNode, undefined);
        tsCodeGen.addText(`}\n`);
    }
    tsCodeGen.addText(`if (typeof __VLS_styleScopedClasses === 'object' && !Array.isArray(__VLS_styleScopedClasses)) {\n`);
    for (const { className, offset } of scopedClasses) {
        tsCodeGen.addText(`__VLS_styleScopedClasses[`);
        writeCodeWithQuotes(className, {
            start: offset,
            end: offset + className.length,
        }, {
            vueTag: 'template',
            capabilities: Object.assign(Object.assign({}, capabilitiesSet.scopedClassName), { displayWithLink: cssScopedClassesSet.has(className) }),
        });
        tsCodeGen.addText(`];\n`);
    }
    tsCodeGen.addText('}\n');
    tsCodeGen.addText(`declare var __VLS_slots:\n`);
    for (const [exp, slot] of slotExps) {
        tsCodeGen.addText(`Record<NonNullable<typeof ${exp}>, typeof ${slot.varName}> &\n`);
    }
    tsCodeGen.addText(`{\n`);
    for (const [name, slot] of slots) {
        writeObjectProperty(name, slot.loc, SourceMaps.Mode.Expand, {
            vueTag: 'template',
            capabilities: capabilitiesSet.slotNameExport,
        });
        tsCodeGen.addText(`: typeof ${slot.varName},\n`);
    }
    tsCodeGen.addText(`};\n`);
    return {
        codeGen: tsCodeGen,
        formatCodeGen: tsFormatCodeGen,
        cssCodeGen: cssCodeGen,
        tagNames: tagResolves,
        attrNames,
        identifiers,
    };
    function collectTags(node) {
        var _a;
        if (node.type === 1 /* ELEMENT */) {
            const patchForNode = getPatchForSlotNode(node);
            if (patchForNode) {
                collectTags(patchForNode);
                return;
            }
            if (!tags[node.tag]) {
                tags[node.tag] = {
                    offsets: [],
                    props: {},
                    events: {},
                };
            }
            const resolvedTag = tags[node.tag];
            resolvedTag.offsets.push(node.loc.start.offset + node.loc.source.indexOf(node.tag)); // start tag
            if (!node.isSelfClosing && sourceLang === 'html') {
                resolvedTag.offsets.push(node.loc.start.offset + node.loc.source.lastIndexOf(node.tag)); // end tag
            }
            for (const prop of node.props) {
                if (prop.type === 7 /* DIRECTIVE */
                    && ((_a = prop.arg) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */
                    && prop.arg.isStatic) {
                    let propName = prop.arg.constType === 3 /* CAN_STRINGIFY */
                        ? prop.arg.content
                        : prop.arg.loc.source;
                    if (prop.modifiers.some(m => m === 'prop' || m === 'attr')) {
                        propName = propName.substring(1);
                    }
                    if (prop.name === 'bind' || prop.name === 'model') {
                        addProp(propName, propName, prop.arg.loc.start.offset);
                    }
                    else if (prop.name === 'on') {
                        addEvent(propName, prop.arg.loc.start.offset);
                    }
                }
                else if (prop.type === 7 /* DIRECTIVE */
                    && !prop.arg
                    && prop.name === 'model') {
                    addProp(getModelValuePropName(node, isVue2), 'v-model', prop.loc.start.offset);
                }
                else if (prop.type === 6 /* ATTRIBUTE */) {
                    addProp(prop.name, prop.name, prop.loc.start.offset);
                }
            }
            for (const childNode of node.children) {
                collectTags(childNode);
            }
            function addProp(propName, argName, offset) {
                if (!resolvedTag.props[propName]) {
                    resolvedTag.props[propName] = {
                        argName,
                        offsets: [],
                    };
                }
                resolvedTag.props[propName].offsets.push(offset);
            }
            function addEvent(eventName, offset) {
                if (!resolvedTag.events[eventName]) {
                    resolvedTag.events[eventName] = {
                        offsets: [],
                    };
                }
                resolvedTag.events[eventName].offsets.push(offset);
            }
        }
        else if (node.type === 9 /* IF */) {
            // v-if / v-else-if / v-else
            for (let i = 0; i < node.branches.length; i++) {
                const branch = node.branches[i];
                for (const childNode of branch.children) {
                    collectTags(childNode);
                }
            }
        }
        else if (node.type === 11 /* FOR */) {
            // v-for
            for (const childNode of node.children) {
                collectTags(childNode);
            }
        }
    }
    function visitNode(node, parentEl) {
        var _a, _b, _c, _d;
        if (node.type === 1 /* ELEMENT */) {
            visitElementNode(node, parentEl);
        }
        else if (node.type === 12 /* TEXT_CALL */) {
            // {{ var }}
            visitNode(node.content, parentEl);
        }
        else if (node.type === 8 /* COMPOUND_EXPRESSION */) {
            // {{ ... }} {{ ... }}
            for (const childNode of node.children) {
                if (typeof childNode === 'object') {
                    visitNode(childNode, parentEl);
                }
            }
        }
        else if (node.type === 5 /* INTERPOLATION */) {
            // {{ ... }}
            const context = node.loc.source.substring(2, node.loc.source.length - 2);
            let start = node.loc.start.offset + 2;
            tsCodeGen.addText(`(`);
            writeInterpolation(context, start, {
                vueTag: 'template',
                capabilities: capabilitiesSet.all,
            }, '', '');
            writeFormatCode(context, start, formatBrackets.curly);
            tsCodeGen.addText(`);\n`);
        }
        else if (node.type === 9 /* IF */) {
            // v-if / v-else-if / v-else
            let originalBlockConditionsLength = blockConditions.length;
            for (let i = 0; i < node.branches.length; i++) {
                const branch = node.branches[i];
                if (i === 0)
                    tsCodeGen.addText('if');
                else if (branch.condition)
                    tsCodeGen.addText('else if');
                else
                    tsCodeGen.addText('else');
                let addedBlockCondition = false;
                if (((_a = branch.condition) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */) {
                    tsCodeGen.addText(` `);
                    writeInterpolation(branch.condition.content, branch.condition.loc.start.offset, {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.all,
                    }, '(', ')');
                    writeFormatCode(branch.condition.content, branch.condition.loc.start.offset, formatBrackets.round);
                    if (allowTypeNarrowingInEventExpressions) {
                        blockConditions.push(branch.condition.content);
                        addedBlockCondition = true;
                    }
                }
                tsCodeGen.addText(` {\n`);
                for (const childNode of branch.children) {
                    visitNode(childNode, parentEl);
                }
                tsCodeGen.addText('}\n');
                if (addedBlockCondition) {
                    blockConditions[blockConditions.length - 1] = `!(${blockConditions[blockConditions.length - 1]})`;
                }
            }
            blockConditions.length = originalBlockConditionsLength;
        }
        else if (node.type === 11 /* FOR */) {
            // v-for
            const { source, value, key, index } = node.parseResult;
            const leftExpressionRange = value ? { start: ((_b = value !== null && value !== void 0 ? value : key) !== null && _b !== void 0 ? _b : index).loc.start.offset, end: ((_c = index !== null && index !== void 0 ? index : key) !== null && _c !== void 0 ? _c : value).loc.end.offset } : undefined;
            const leftExpressionText = leftExpressionRange ? node.loc.source.substring(leftExpressionRange.start - node.loc.start.offset, leftExpressionRange.end - node.loc.start.offset) : undefined;
            const forBlockVars = [];
            tsCodeGen.addText(`for (const [`);
            if (leftExpressionRange && leftExpressionText) {
                const collentAst = ts.createSourceFile('/foo.ts', `const [${leftExpressionText}]`, ts.ScriptTarget.ESNext);
                (0, transform_1.colletVars)(ts, collentAst, forBlockVars);
                for (const varName of forBlockVars)
                    localVars[varName] = ((_d = localVars[varName]) !== null && _d !== void 0 ? _d : 0) + 1;
                writeCode(leftExpressionText, leftExpressionRange, SourceMaps.Mode.Offset, {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.all,
                });
                writeFormatCode(leftExpressionText, leftExpressionRange.start, formatBrackets.square);
            }
            tsCodeGen.addText(`] of __VLS_types.getVforSourceType`);
            if (source.type === 4 /* SIMPLE_EXPRESSION */) {
                writeInterpolation(source.content, source.loc.start.offset, {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.all,
                }, '(', ')');
                writeFormatCode(source.content, source.loc.start.offset, formatBrackets.round);
            }
            tsCodeGen.addText(`) {\n`);
            for (const childNode of node.children) {
                visitNode(childNode, parentEl);
            }
            tsCodeGen.addText('}\n');
            for (const varName of forBlockVars)
                localVars[varName]--;
        }
        else if (node.type === 2 /* TEXT */) {
            // not needed progress
        }
        else if (node.type === 3 /* COMMENT */) {
            // not needed progress
        }
        else {
            tsCodeGen.addText(`// Unprocessed node type: ${node.type} json: ${JSON.stringify(node.loc)}\n`);
        }
    }
    ;
    function visitElementNode(node, parentEl) {
        var _a;
        const patchForNode = getPatchForSlotNode(node);
        if (patchForNode) {
            visitNode(patchForNode, parentEl);
            return;
        }
        if (node.tag !== 'template') {
            parentEl = node;
        }
        tsCodeGen.addText(`{\n`);
        {
            const tagText = experimentalRuntimeMode === 'runtime-dom' && ((0, shared_1.isHTMLTag)(node.tag) || (0, shared_1.isSVGTag)(node.tag)) ? node.tag : tagResolves[node.tag].rawComponent;
            const fullTagStart = tsCodeGen.getText().length;
            tsCodeGen.addText(`<`);
            writeCode(tagText, {
                start: node.loc.start.offset + node.loc.source.indexOf(node.tag),
                end: node.loc.start.offset + node.loc.source.indexOf(node.tag) + node.tag.length,
            }, SourceMaps.Mode.Totally, {
                vueTag: 'template',
                capabilities: capabilitiesSet.diagnosticOnly,
            });
            tsCodeGen.addText(` `);
            const { hasRemainStyleOrClass } = writeProps(node, false, 'props');
            tsCodeGen.addText(`/>`);
            // fix https://github.com/johnsoncodehk/volar/issues/705#issuecomment-974773353
            let startTagEnd;
            if (node.loc.source.endsWith('/>')) {
                startTagEnd = node.loc.end.offset;
            }
            else if (node.children.length) {
                startTagEnd = node.loc.start.offset + node.loc.source.substring(0, node.children[0].loc.start.offset - node.loc.start.offset).lastIndexOf('>') + 1;
            }
            else {
                startTagEnd = node.loc.start.offset + node.loc.source.substring(0, node.loc.source.lastIndexOf('</')).lastIndexOf('>') + 1;
            }
            addMapping(tsCodeGen, {
                sourceRange: {
                    start: node.loc.start.offset,
                    end: startTagEnd,
                },
                mappedRange: {
                    start: fullTagStart,
                    end: tsCodeGen.getText().length,
                },
                mode: SourceMaps.Mode.Totally,
                data: {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.diagnosticOnly,
                },
            });
            tsCodeGen.addText(`\n`);
            if (hasRemainStyleOrClass) {
                tsCodeGen.addText(`<${tagText} `);
                writeProps(node, true, 'props');
                tsCodeGen.addText(`/>\n`);
            }
            let slotBlockVars;
            writeInlineCss(node);
            if (parentEl) {
                slotBlockVars = [];
                writeImportSlots(node, parentEl, slotBlockVars);
                for (const varName of slotBlockVars)
                    localVars[varName] = ((_a = localVars[varName]) !== null && _a !== void 0 ? _a : 0) + 1;
            }
            writeDirectives(node);
            writeElReferences(node); // <el ref="foo" />
            if (cssScopedClasses.length)
                writeClassScopeds(node);
            writeEvents(node);
            writeSlots(node);
            for (const childNode of node.children) {
                visitNode(childNode, parentEl);
            }
            if (slotBlockVars) {
                for (const varName of slotBlockVars)
                    localVars[varName]--;
            }
        }
        tsCodeGen.addText(`}\n`);
        function writeEvents(node) {
            var _a, _b;
            const varComponentInstance = `__VLS_${elementIndex++}`;
            let writedInstance = false;
            for (const prop of node.props) {
                if (prop.type === 7 /* DIRECTIVE */
                    && prop.name === 'on'
                    && ((_a = prop.arg) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */) {
                    const transformResult = CompilerDOM.transformOn(prop, node, exports.transformContext);
                    for (const prop_2 of transformResult.props) {
                        tryWriteInstance();
                        const varInstanceProps = `__VLS_${elementIndex++}`;
                        tsCodeGen.addText(`type ${varInstanceProps} = typeof ${varComponentInstance} extends { $props: infer Props } ? Props & Omit<__VLS_types.GlobalAttrs, keyof Props> & Record<string, unknown> : typeof ${tagResolves[node.tag].rawComponent} & Record<string, unknown>;\n`);
                        tsCodeGen.addText(`const __VLS_${elementIndex++}: {\n`);
                        tsCodeGen.addText(`'${prop.arg.loc.source}': __VLS_types.FillingEventArg<\n`);
                        {
                            tsCodeGen.addText(`__VLS_types.FirstFunction<\n`);
                            {
                                tsCodeGen.addText(`__VLS_types.EmitEvent<typeof ${tagResolves[node.tag].rawComponent}, '${prop.arg.loc.source}'>,\n`);
                                {
                                    const key_2 = (0, shared_1.camelize)('on-' + prop.arg.loc.source); // onClickOutside
                                    const key_3 = 'on' + prop.arg.loc.source[0].toUpperCase() + prop.arg.loc.source.substring(1); // onClick-outside
                                    tsCodeGen.addText(`${varInstanceProps}[`);
                                    writeCodeWithQuotes(key_2, [{ start: prop.arg.loc.start.offset, end: prop.arg.loc.end.offset }], {
                                        vueTag: 'template',
                                        capabilities: capabilitiesSet.attrReference,
                                        normalizeNewName(newName) {
                                            return (0, shared_1.camelize)('on-' + newName);
                                        },
                                        applyNewName(oldName, newName) {
                                            const hName = (0, shared_1.hyphenate)(newName);
                                            if ((0, shared_1.hyphenate)(newName).startsWith('on-')) {
                                                return (0, shared_1.camelize)(hName.slice('on-'.length));
                                            }
                                            return newName;
                                        },
                                    });
                                    tsCodeGen.addText(`],\n`);
                                    if (key_3 !== key_2) {
                                        tsCodeGen.addText(`${varInstanceProps}[`);
                                        writeCodeWithQuotes(key_3, [{ start: prop.arg.loc.start.offset, end: prop.arg.loc.end.offset }], {
                                            vueTag: 'template',
                                            capabilities: capabilitiesSet.attrReference,
                                            normalizeNewName(newName) {
                                                return 'on' + newName[0].toUpperCase() + newName.substring(1);
                                            },
                                            applyNewName(oldName, newName) {
                                                const hName = (0, shared_1.hyphenate)(newName);
                                                if ((0, shared_1.hyphenate)(newName).startsWith('on-')) {
                                                    return (0, shared_1.camelize)(hName.slice('on-'.length));
                                                }
                                                return newName;
                                            },
                                        });
                                        tsCodeGen.addText(`],\n`);
                                    }
                                }
                                tsCodeGen.addText(`typeof ${varComponentInstance} extends { $emit: infer Emit } ? __VLS_types.EmitEvent2<Emit, '${prop.arg.loc.source}'> : unknown,\n`);
                            }
                            tsCodeGen.addText(`>\n`);
                        }
                        tsCodeGen.addText(`>\n`);
                        tsCodeGen.addText(`} = {\n`);
                        {
                            writeObjectProperty(prop.arg.loc.source, {
                                start: prop.arg.loc.start.offset,
                                end: prop.arg.loc.end.offset,
                            }, SourceMaps.Mode.Offset, {
                                vueTag: 'template',
                                capabilities: capabilitiesSet.event,
                            });
                            tsCodeGen.addText(`: `);
                            appendExpressionNode(prop, prop_2.value);
                        }
                        tsCodeGen.addText(`};\n`);
                    }
                }
                else if (prop.type === 7 /* DIRECTIVE */
                    && prop.name === 'on'
                    && ((_b = prop.exp) === null || _b === void 0 ? void 0 : _b.type) === 4 /* SIMPLE_EXPRESSION */) {
                    // for vue 2 nameless event
                    // https://github.com/johnsoncodehk/vue-tsc/issues/67
                    writeInterpolation(prop.exp.content, prop.exp.loc.start.offset, {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.all,
                    }, '$event => {(', ')}');
                    writeFormatCode(prop.exp.content, prop.exp.loc.start.offset, formatBrackets.round);
                    tsCodeGen.addText(`;\n`);
                }
                function appendExpressionNode(prop, jsChildNode) {
                    var _a;
                    if (((_a = prop.exp) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */) {
                        if (jsChildNode.type === 4 /* SIMPLE_EXPRESSION */) {
                            writeInterpolation(prop.exp.content, prop.exp.loc.start.offset, {
                                vueTag: 'template',
                                capabilities: capabilitiesSet.all,
                            }, '(', ')');
                            writeFormatCode(prop.exp.content, prop.exp.loc.start.offset, formatBrackets.round);
                        }
                        else if (jsChildNode.type === 8 /* COMPOUND_EXPRESSION */) {
                            const _exp = prop.exp;
                            const expIndex = jsChildNode.children.findIndex(child => typeof child === 'object' && child.type === 4 /* SIMPLE_EXPRESSION */ && child.content === _exp.content);
                            const expNode = jsChildNode.children[expIndex];
                            let prefix = jsChildNode.children.filter((child, i) => typeof child === 'string' && i < expIndex).map(child => child).join('');
                            let suffix = jsChildNode.children.filter((child, i) => typeof child === 'string' && i > expIndex).map(child => child).join('');
                            if (prefix && blockConditions.length) {
                                prefix = prefix.replace('(', '{ ');
                                suffix = suffix.replace(')', '} ');
                                prefix += '\n';
                                for (const blockCondition of blockConditions) {
                                    prefix += `if (!(${blockCondition})) return;\n`;
                                }
                            }
                            writeInterpolation(expNode.content, expNode.loc.start.offset, {
                                vueTag: 'template',
                                capabilities: capabilitiesSet.all,
                            }, prefix, suffix);
                            writeFormatCode(expNode.content, expNode.loc.start.offset, formatBrackets.round);
                        }
                    }
                    else {
                        tsCodeGen.addText(`undefined`);
                    }
                }
            }
            function tryWriteInstance() {
                if (writedInstance)
                    return;
                tsCodeGen.addText(`const ${varComponentInstance} = new ${tagResolves[node.tag].rawComponent}({ `);
                writeProps(node, false, 'slots');
                tsCodeGen.addText(`});\n`);
                writedInstance = true;
            }
        }
    }
    function writeProps(node, forRemainStyleOrClass, mode) {
        var _a, _b, _c, _d, _e;
        let styleCount = 0;
        let classCount = 0;
        for (const prop of node.props) {
            if (prop.type === 7 /* DIRECTIVE */
                && (prop.name === 'bind' || prop.name === 'model')
                && (prop.name === 'model' || ((_a = prop.arg) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */)
                && (!prop.exp || prop.exp.type === 4 /* SIMPLE_EXPRESSION */)) {
                const isStatic = !prop.arg || (prop.arg.type === 4 /* SIMPLE_EXPRESSION */ && prop.arg.isStatic);
                let propName_1 = ((_b = prop.arg) === null || _b === void 0 ? void 0 : _b.type) === 4 /* SIMPLE_EXPRESSION */
                    ? prop.arg.constType === 3 /* CAN_STRINGIFY */
                        ? prop.arg.content
                        : prop.arg.loc.source
                    : getModelValuePropName(node, isVue2);
                if (prop.modifiers.some(m => m === 'prop' || m === 'attr')) {
                    propName_1 = propName_1.substring(1);
                }
                const propName_2 = !isStatic ? propName_1 : (0, shared_1.hyphenate)(propName_1) === propName_1 ? (0, shared_1.camelize)(propName_1) : propName_1;
                if (forRemainStyleOrClass && propName_2 !== 'style' && propName_2 !== 'class')
                    continue;
                if (propName_2 === 'style' || propName_2 === 'class') {
                    const index = propName_2 === 'style' ? styleCount++ : classCount++;
                    if (index >= 1 !== forRemainStyleOrClass)
                        continue;
                }
                if (((_c = prop.arg) === null || _c === void 0 ? void 0 : _c.type) === 4 /* SIMPLE_EXPRESSION */) {
                    attrNames.add(prop.arg.content);
                }
                // camelize name
                writePropStart(isStatic);
                const diagStart = tsCodeGen.getText().length;
                if (!prop.arg) {
                    writePropName(propName_1, isStatic, {
                        start: prop.loc.start.offset,
                        end: prop.loc.start.offset + 'v-model'.length,
                    }, {
                        vueTag: 'template',
                        capabilities: getCaps(capabilitiesSet.attr),
                    });
                }
                else if (((_d = prop.exp) === null || _d === void 0 ? void 0 : _d.constType) === 3 /* CAN_STRINGIFY */) {
                    writePropName(propName_2, isStatic, {
                        start: prop.arg.loc.start.offset,
                        end: prop.arg.loc.start.offset + propName_1.length, // patch style attr
                    }, {
                        vueTag: 'template',
                        capabilities: getCaps(capabilitiesSet.attr),
                        normalizeNewName: shared_1.camelize,
                        applyNewName: keepHyphenateName,
                    });
                }
                else {
                    writePropName(propName_2, isStatic, {
                        start: prop.arg.loc.start.offset,
                        end: prop.arg.loc.end.offset,
                    }, {
                        vueTag: 'template',
                        capabilities: getCaps(capabilitiesSet.attr),
                        normalizeNewName: shared_1.camelize,
                        applyNewName: keepHyphenateName,
                    });
                }
                writePropValuePrefix(isStatic);
                if (prop.exp && !(prop.exp.constType === 3 /* CAN_STRINGIFY */)) { // style='z-index: 2' will compile to {'z-index':'2'}
                    writeInterpolation(prop.exp.loc.source, prop.exp.loc.start.offset, {
                        vueTag: 'template',
                        capabilities: getCaps(capabilitiesSet.all),
                    }, '(', ')');
                    const fb = getFormatBrackets(formatBrackets.round);
                    if (fb) {
                        writeFormatCode(prop.exp.loc.source, prop.exp.loc.start.offset, fb);
                    }
                }
                else {
                    tsCodeGen.addText('undefined');
                }
                writePropValueSuffix(isStatic);
                addMapping(tsCodeGen, {
                    sourceRange: {
                        start: prop.loc.start.offset,
                        end: prop.loc.end.offset,
                    },
                    mappedRange: {
                        start: diagStart,
                        end: tsCodeGen.getText().length,
                    },
                    mode: SourceMaps.Mode.Totally,
                    data: {
                        vueTag: 'template',
                        capabilities: getCaps(capabilitiesSet.diagnosticOnly),
                    },
                });
                writePropEnd(isStatic);
                // original name
                if (prop.arg && propName_1 !== propName_2) {
                    writePropStart(isStatic);
                    writePropName(propName_1, isStatic, {
                        start: prop.arg.loc.start.offset,
                        end: prop.arg.loc.end.offset,
                    }, {
                        vueTag: 'template',
                        capabilities: getCaps(capabilitiesSet.attr),
                        normalizeNewName: shared_1.camelize,
                        applyNewName: keepHyphenateName,
                    });
                    writePropValuePrefix(isStatic);
                    if (prop.exp) {
                        writeInterpolation(prop.exp.loc.source, undefined, undefined, '(', ')');
                    }
                    else {
                        tsCodeGen.addText('undefined');
                    }
                    writePropValueSuffix(isStatic);
                    writePropEnd(isStatic);
                }
            }
            else if (prop.type === 6 /* ATTRIBUTE */) {
                const propName = (0, shared_1.hyphenate)(prop.name) === prop.name ? (0, shared_1.camelize)(prop.name) : prop.name;
                const propName2 = prop.name;
                if (forRemainStyleOrClass && propName !== 'style' && propName !== 'class')
                    continue;
                if (propName === 'style' || propName === 'class') {
                    const index = propName === 'style' ? styleCount++ : classCount++;
                    if (index >= 1 !== forRemainStyleOrClass)
                        continue;
                }
                attrNames.add(prop.name);
                // camelize name
                writePropStart(true);
                const diagStart = tsCodeGen.getText().length;
                writePropName(propName, true, {
                    start: prop.loc.start.offset,
                    end: prop.loc.start.offset + propName2.length,
                }, {
                    vueTag: 'template',
                    capabilities: getCaps(capabilitiesSet.attr),
                    normalizeNewName: shared_1.camelize,
                    applyNewName: keepHyphenateName,
                });
                writePropValuePrefix(true);
                if (prop.value) {
                    writeAttrValue(prop.value);
                }
                else {
                    tsCodeGen.addText('true');
                }
                writePropValueSuffix(true);
                writePropEnd(true);
                const diagEnd = tsCodeGen.getText().length;
                addMapping(tsCodeGen, {
                    sourceRange: {
                        start: prop.loc.start.offset,
                        end: prop.loc.end.offset,
                    },
                    mappedRange: {
                        start: diagStart,
                        end: diagEnd,
                    },
                    mode: SourceMaps.Mode.Totally,
                    data: {
                        vueTag: 'template',
                        capabilities: getCaps(capabilitiesSet.diagnosticOnly),
                    },
                });
                // original name
                if (propName2 !== propName) {
                    writePropStart(true);
                    writePropName(propName2, true, {
                        start: prop.loc.start.offset,
                        end: prop.loc.start.offset + propName2.length,
                    }, {
                        vueTag: 'template',
                        capabilities: getCaps(capabilitiesSet.attr),
                        normalizeNewName: shared_1.camelize,
                        applyNewName: keepHyphenateName,
                    });
                    writePropValuePrefix(true);
                    if (prop.value) {
                        writeAttrValue(prop.value);
                    }
                    else {
                        tsCodeGen.addText('true');
                    }
                    writePropValueSuffix(true);
                    writePropEnd(true);
                }
            }
            else if (prop.type === 7 /* DIRECTIVE */
                && prop.name === 'bind'
                && !prop.arg
                && ((_e = prop.exp) === null || _e === void 0 ? void 0 : _e.type) === 4 /* SIMPLE_EXPRESSION */) {
                if (forRemainStyleOrClass) {
                    continue;
                }
                if (mode === 'props')
                    tsCodeGen.addText('{...');
                else
                    tsCodeGen.addText('...');
                writeInterpolation(prop.exp.content, prop.exp.loc.start.offset, {
                    vueTag: 'template',
                    capabilities: getCaps(capabilitiesSet.all),
                }, '(', ')');
                const fb = getFormatBrackets(formatBrackets.round);
                if (fb) {
                    writeFormatCode(prop.exp.content, prop.exp.loc.start.offset, fb);
                }
                if (mode === 'props')
                    tsCodeGen.addText('} ');
                else
                    tsCodeGen.addText(', ');
            }
            else {
                if (forRemainStyleOrClass) {
                    continue;
                }
                // comment this line to avoid affecting comments in prop expressions
                // tsCodeGen.addText("/* " + [prop.type, prop.name, prop.arg?.loc.source, prop.exp?.loc.source, prop.loc.source].join(", ") + " */ ");
            }
        }
        return { hasRemainStyleOrClass: styleCount >= 2 || classCount >= 2 };
        function writePropName(name, isStatic, sourceRange, data) {
            if (mode === 'props' && isStatic) {
                writeCode(name, sourceRange, SourceMaps.Mode.Offset, data);
            }
            else {
                writeObjectProperty(name, sourceRange, SourceMaps.Mode.Offset, data);
            }
        }
        function writePropValuePrefix(isStatic) {
            if (mode === 'props' && isStatic) {
                tsCodeGen.addText('={');
            }
            else {
                tsCodeGen.addText(': (');
            }
        }
        function writePropValueSuffix(isStatic) {
            if (mode === 'props' && isStatic) {
                tsCodeGen.addText('}');
            }
            else {
                tsCodeGen.addText(')');
            }
        }
        function writePropStart(isStatic) {
            if (mode === 'props' && !isStatic) {
                tsCodeGen.addText('{...{');
            }
        }
        function writePropEnd(isStatic) {
            if (mode === 'props' && isStatic) {
                tsCodeGen.addText(' ');
            }
            else if (mode === 'props' && !isStatic) {
                tsCodeGen.addText('}} ');
            }
            else {
                tsCodeGen.addText(', ');
            }
        }
        function getCaps(caps) {
            if (mode === 'props') {
                return caps;
            }
            else {
                return {
                    references: caps.references,
                    rename: caps.rename,
                };
            }
        }
        function getFormatBrackets(b) {
            if (mode === 'props') {
                return b;
            }
            else {
                return undefined;
            }
        }
        function writeAttrValue(attrNode) {
            tsCodeGen.addText('"');
            let start = attrNode.loc.start.offset;
            let end = attrNode.loc.end.offset;
            if (end - start > attrNode.content.length) {
                start++;
                end--;
            }
            writeCode(toUnicode(attrNode.content), { start, end }, SourceMaps.Mode.Offset, {
                vueTag: 'template',
                capabilities: getCaps(capabilitiesSet.all)
            });
            tsCodeGen.addText('"');
        }
    }
    function writeInlineCss(node) {
        var _a, _b, _c;
        for (const prop of node.props) {
            if (prop.type === 7 /* DIRECTIVE */
                && prop.name === 'bind'
                && ((_a = prop.arg) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */
                && ((_b = prop.exp) === null || _b === void 0 ? void 0 : _b.type) === 4 /* SIMPLE_EXPRESSION */
                && prop.arg.content === 'style'
                && prop.exp.constType === 3 /* CAN_STRINGIFY */) {
                const endCrt = prop.arg.loc.source[prop.arg.loc.source.length - 1]; // " | '
                const start = prop.arg.loc.source.indexOf(endCrt) + 1;
                const end = prop.arg.loc.source.lastIndexOf(endCrt);
                const content = prop.arg.loc.source.substring(start, end);
                const sourceRange = {
                    start: prop.arg.loc.start.offset + start,
                    end: prop.arg.loc.start.offset + end,
                };
                const newStart = (_c = htmlToTemplate(sourceRange.start, sourceRange.end)) === null || _c === void 0 ? void 0 : _c.start;
                if (newStart === undefined)
                    continue;
                const offset = newStart - sourceRange.start;
                sourceRange.start += offset;
                sourceRange.end += offset;
                cssCodeGen.addText(`${node.tag} { `);
                cssCodeGen.addCode(content, sourceRange, SourceMaps.Mode.Offset, {
                    vueTag: 'template',
                    capabilities: {
                        basic: true,
                        references: true,
                        definitions: true,
                        diagnostic: true,
                        rename: true,
                        completion: true,
                        semanticTokens: true,
                    },
                });
                cssCodeGen.addText(` }\n`);
            }
        }
    }
    function writeImportSlots(node, parentEl, slotBlockVars) {
        var _a, _b;
        for (const prop of node.props) {
            if (prop.type === 7 /* DIRECTIVE */
                && prop.name === 'slot') {
                const varComponentInstance = `__VLS_${elementIndex++}`;
                const varSlots = `__VLS_${elementIndex++}`;
                tsCodeGen.addText(`const ${varComponentInstance} = new ${tagResolves[parentEl.tag].slotsComponent}({ `);
                writeProps(parentEl, false, 'slots');
                tsCodeGen.addText(`});\n`);
                tsCodeGen.addText(`declare const ${varSlots}: typeof ${tagResolves[parentEl.tag].slots} & __VLS_types.ScriptSlots<typeof ${varComponentInstance}>;\n`);
                if (((_a = prop.exp) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */) {
                    tsCodeGen.addText(`const `);
                    const collentAst = ts.createSourceFile('/foo.ts', `const ${prop.exp.content}`, ts.ScriptTarget.ESNext);
                    (0, transform_1.colletVars)(ts, collentAst, slotBlockVars);
                    writeCode(prop.exp.content, {
                        start: prop.exp.loc.start.offset,
                        end: prop.exp.loc.end.offset,
                    }, SourceMaps.Mode.Offset, {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.all,
                    });
                    writeFormatCode(prop.exp.content, prop.exp.loc.start.offset, formatBrackets.round);
                    tsCodeGen.addText(` = `);
                }
                let slotName = 'default';
                let isStatic = true;
                if (((_b = prop.arg) === null || _b === void 0 ? void 0 : _b.type) === 4 /* SIMPLE_EXPRESSION */ && prop.arg.content !== '') {
                    isStatic = prop.arg.isStatic;
                    slotName = prop.arg.content;
                }
                const diagStart = tsCodeGen.getText().length;
                tsCodeGen.addText(varSlots);
                const argRange = prop.arg
                    ? {
                        start: prop.arg.loc.start.offset,
                        end: prop.arg.loc.end.offset,
                    } : {
                    start: prop.loc.start.offset,
                    end: prop.loc.start.offset + prop.loc.source.split('=')[0].length,
                };
                if (isStatic) {
                    writePropertyAccess(slotName, argRange, {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.slotName,
                    }, false);
                }
                else {
                    tsCodeGen.addText(`[`);
                    writeInterpolation(slotName, argRange.start + 1, {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.all,
                    }, '', '');
                    tsCodeGen.addText(`]`);
                }
                const diagEnd = tsCodeGen.getText().length;
                addMapping(tsCodeGen, {
                    mappedRange: {
                        start: diagStart,
                        end: diagEnd,
                    },
                    sourceRange: argRange,
                    mode: SourceMaps.Mode.Totally,
                    data: {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.diagnosticOnly,
                    },
                });
                tsCodeGen.addText(`;\n`);
            }
        }
    }
    function writeDirectives(node) {
        var _a;
        for (const prop of node.props) {
            if (prop.type === 7 /* DIRECTIVE */
                && prop.name !== 'slot'
                && prop.name !== 'on'
                && prop.name !== 'model'
                && prop.name !== 'bind') {
                const diagStart = tsCodeGen.getText().length;
                tsCodeGen.addText(`__VLS_types.directiveFunction(__VLS_ctx.`);
                writeCode((0, shared_1.camelize)('v-' + prop.name), {
                    start: prop.loc.start.offset,
                    end: prop.loc.start.offset + 'v-'.length + prop.name.length,
                }, SourceMaps.Mode.Offset, {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.noDiagnostic,
                    normalizeNewName: shared_1.camelize,
                    applyNewName: keepHyphenateName,
                });
                identifiers.add((0, shared_1.camelize)('v-' + prop.name));
                tsCodeGen.addText(`)(`);
                if (((_a = prop.exp) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */) {
                    writeInterpolation(prop.exp.content, prop.exp.loc.start.offset, {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.all,
                    }, '(', ')');
                    writeFormatCode(prop.exp.content, prop.exp.loc.start.offset, formatBrackets.round);
                }
                tsCodeGen.addText(`)`);
                addMapping(tsCodeGen, {
                    sourceRange: {
                        start: prop.loc.start.offset,
                        end: prop.loc.end.offset,
                    },
                    mappedRange: {
                        start: diagStart,
                        end: tsCodeGen.getText().length,
                    },
                    mode: SourceMaps.Mode.Totally,
                    data: {
                        vueTag: 'template',
                        capabilities: capabilitiesSet.diagnosticOnly,
                    },
                });
                tsCodeGen.addText(`;\n`);
            }
        }
    }
    function writeElReferences(node) {
        for (const prop of node.props) {
            if (prop.type === 6 /* ATTRIBUTE */
                && prop.name === 'ref'
                && prop.value) {
                tsCodeGen.addText(`// @ts-ignore\n`);
                writeInterpolation(prop.value.content, prop.value.loc.start.offset + 1, {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.refAttr,
                }, '(', ')');
                tsCodeGen.addText(`;\n`);
            }
        }
    }
    function writeClassScopeds(node) {
        var _a, _b;
        for (const prop of node.props) {
            if (prop.type === 6 /* ATTRIBUTE */
                && prop.name === 'class'
                && prop.value) {
                let startOffset = prop.value.loc.start.offset;
                let tempClassName = '';
                for (const char of (prop.value.loc.source + ' ')) {
                    if (char.trim() === '' || char === '"' || char === "'") {
                        if (tempClassName !== '') {
                            scopedClasses.push({ className: tempClassName, offset: startOffset });
                            startOffset += tempClassName.length;
                            tempClassName = '';
                        }
                        startOffset += char.length;
                    }
                    else {
                        tempClassName += char;
                    }
                }
            }
            else if (prop.type === 7 /* DIRECTIVE */
                && ((_a = prop.arg) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */
                && ((_b = prop.exp) === null || _b === void 0 ? void 0 : _b.type) === 4 /* SIMPLE_EXPRESSION */
                && prop.arg.content === 'class') {
                tsCodeGen.addText(`__VLS_styleScopedClasses = (`);
                writeCode(prop.exp.content, {
                    start: prop.exp.loc.start.offset,
                    end: prop.exp.loc.end.offset,
                }, SourceMaps.Mode.Offset, {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.scopedClassName,
                });
                tsCodeGen.addText(`);\n`);
            }
        }
    }
    function writeSlots(node) {
        var _a, _b, _c;
        if (node.tag !== 'slot')
            return;
        const varDefaultBind = `__VLS_${elementIndex++}`;
        const varBinds = `__VLS_${elementIndex++}`;
        const varSlot = `__VLS_${elementIndex++}`;
        const slotName = getSlotName();
        const slotNameExp = getSlotNameExp();
        let hasDefaultBind = false;
        for (const prop of node.props) {
            if (prop.type === 7 /* DIRECTIVE */
                && !prop.arg
                && ((_a = prop.exp) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */) {
                hasDefaultBind = true;
                tsCodeGen.addText(`const ${varDefaultBind} = `);
                writeInterpolation(prop.exp.content, prop.exp.loc.start.offset, {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.attrReference,
                }, '(', ')');
                tsCodeGen.addText(`;\n`);
                break;
            }
        }
        tsCodeGen.addText(`const ${varBinds} = {\n`);
        for (const prop of node.props) {
            if (prop.type === 7 /* DIRECTIVE */
                && ((_b = prop.arg) === null || _b === void 0 ? void 0 : _b.type) === 4 /* SIMPLE_EXPRESSION */
                && ((_c = prop.exp) === null || _c === void 0 ? void 0 : _c.type) === 4 /* SIMPLE_EXPRESSION */
                && prop.arg.content !== 'name') {
                writeObjectProperty(prop.arg.content, {
                    start: prop.arg.loc.start.offset,
                    end: prop.arg.loc.end.offset,
                }, SourceMaps.Mode.Offset, {
                    vueTag: 'template',
                    normalizeNewName: shared_1.camelize,
                    applyNewName: keepHyphenateName,
                    capabilities: capabilitiesSet.attrReference,
                });
                tsCodeGen.addText(`: `);
                writeInterpolation(prop.exp.content, prop.exp.loc.start.offset, {
                    vueTag: 'template',
                    capabilities: capabilitiesSet.attrReference,
                }, '(', ')');
                tsCodeGen.addText(`,\n`);
            }
            else if (prop.type === 6 /* ATTRIBUTE */
                && prop.name !== 'name' // slot name
            ) {
                const propValue = prop.value !== undefined ? `"${toUnicode(prop.value.content)}"` : 'true';
                writeObjectProperty(prop.name, {
                    start: prop.loc.start.offset,
                    end: prop.loc.start.offset + prop.name.length
                }, SourceMaps.Mode.Offset, {
                    vueTag: 'template',
                    normalizeNewName: shared_1.camelize,
                    applyNewName: keepHyphenateName,
                    capabilities: capabilitiesSet.attr,
                });
                tsCodeGen.addText(`: (`);
                tsCodeGen.addText(propValue);
                tsCodeGen.addText(`),\n`);
            }
        }
        tsCodeGen.addText(`};\n`);
        if (hasDefaultBind) {
            tsCodeGen.addText(`var ${varSlot}!: typeof ${varDefaultBind} & typeof ${varBinds};\n`);
        }
        else {
            tsCodeGen.addText(`var ${varSlot}!: typeof ${varBinds};\n`);
        }
        if (slotNameExp) {
            const varSlotExp = `__VLS_${elementIndex++}`;
            const varSlotExp2 = `__VLS_${elementIndex++}`;
            tsCodeGen.addText(`const ${varSlotExp} = ${slotNameExp};\n`);
            tsCodeGen.addText(`var ${varSlotExp2}!: typeof ${varSlotExp};\n`);
            slotExps.set(varSlotExp2, {
                varName: varSlot,
                loc: {
                    start: node.loc.start.offset + node.loc.source.indexOf(node.tag),
                    end: node.loc.start.offset + node.loc.source.indexOf(node.tag) + node.tag.length,
                },
            });
        }
        else {
            slots.set(slotName, {
                varName: varSlot,
                loc: {
                    start: node.loc.start.offset + node.loc.source.indexOf(node.tag),
                    end: node.loc.start.offset + node.loc.source.indexOf(node.tag) + node.tag.length,
                },
            });
        }
        function getSlotName() {
            for (const prop2 of node.props) {
                if (prop2.name === 'name' && prop2.type === 6 /* ATTRIBUTE */ && prop2.value) {
                    if (prop2.value.content) {
                        return prop2.value.content;
                    }
                }
            }
            return 'default';
        }
        function getSlotNameExp() {
            var _a, _b;
            for (const prop2 of node.props) {
                if (prop2.type === 7 /* DIRECTIVE */ && prop2.name === 'bind' && ((_a = prop2.arg) === null || _a === void 0 ? void 0 : _a.type) === 4 /* SIMPLE_EXPRESSION */ && prop2.arg.content === 'name') {
                    if (((_b = prop2.exp) === null || _b === void 0 ? void 0 : _b.type) === 4 /* SIMPLE_EXPRESSION */) {
                        return prop2.exp.content;
                    }
                    else {
                        return `'default'`;
                    }
                }
            }
        }
    }
    function writeObjectProperty2(mapCode, sourceRanges, data) {
        const sourceRange = sourceRanges[0];
        const mode = writeObjectProperty(mapCode, sourceRange, SourceMaps.Mode.Offset, data);
        for (let i = 1; i < sourceRanges.length; i++) {
            const sourceRange = sourceRanges[i];
            if (mode === 1) {
                addMapping(tsCodeGen, {
                    sourceRange,
                    mappedRange: {
                        start: tsCodeGen.getText().length - mapCode.length,
                        end: tsCodeGen.getText().length,
                    },
                    mode: SourceMaps.Mode.Offset,
                    data,
                });
            }
            else if (mode === 2) {
                addMapping(tsCodeGen, {
                    sourceRange,
                    mappedRange: {
                        start: tsCodeGen.getText().length - `'${mapCode}'`.length,
                        end: tsCodeGen.getText().length - `'`.length,
                    },
                    mode: SourceMaps.Mode.Offset,
                    additional: [
                        {
                            sourceRange,
                            mappedRange: {
                                start: tsCodeGen.getText().length - `'${mapCode}'`.length,
                                end: tsCodeGen.getText().length,
                            },
                            mode: SourceMaps.Mode.Totally,
                        }
                    ],
                    data,
                });
            }
        }
    }
    function writeObjectProperty(mapCode, sourceRange, mapMode, data) {
        if (validTsVar.test(mapCode)) {
            writeCode(mapCode, sourceRange, mapMode, data);
            return 1;
        }
        else if (mapCode.startsWith('[') && mapCode.endsWith(']')) {
            writeInterpolation(mapCode, sourceRange.start, data, '', '');
            return 1;
        }
        else {
            writeCodeWithQuotes(mapCode, sourceRange, data);
            return 2;
        }
    }
    function writePropertyAccess2(mapCode, sourceRanges, data) {
        const sourceRange = sourceRanges[0];
        const mode = writePropertyAccess(mapCode, sourceRange, data);
        for (let i = 1; i < sourceRanges.length; i++) {
            const sourceRange = sourceRanges[i];
            if (mode === 1 || mode === 2) {
                addMapping(tsCodeGen, {
                    sourceRange,
                    mappedRange: {
                        start: tsCodeGen.getText().length - mapCode.length,
                        end: tsCodeGen.getText().length,
                    },
                    mode: sourceRange.end - sourceRange.start === mapCode.length ? SourceMaps.Mode.Offset : SourceMaps.Mode.Expand,
                    data,
                });
            }
            else if (mode === 3) {
                addMapping(tsCodeGen, {
                    sourceRange,
                    mappedRange: {
                        start: tsCodeGen.getText().length - `['${mapCode}']`.length,
                        end: tsCodeGen.getText().length - `']`.length,
                    },
                    mode: SourceMaps.Mode.Offset,
                    additional: [
                        {
                            sourceRange,
                            mappedRange: {
                                start: tsCodeGen.getText().length - `'${mapCode}']`.length,
                                end: tsCodeGen.getText().length - `]`.length,
                            },
                            mode: SourceMaps.Mode.Totally,
                        }
                    ],
                    data,
                });
            }
        }
    }
    function writePropertyAccess(mapCode, sourceRange, data, checkValid = true) {
        if (checkValid && validTsVar.test(mapCode)) {
            tsCodeGen.addText(`.`);
            if (sourceRange.end - sourceRange.start === mapCode.length) {
                writeCode(mapCode, sourceRange, SourceMaps.Mode.Offset, data);
            }
            else {
                writeCode(mapCode, sourceRange, SourceMaps.Mode.Expand, data);
            }
            return 1;
        }
        else if (mapCode.startsWith('[') && mapCode.endsWith(']')) {
            writeCode(mapCode, sourceRange, SourceMaps.Mode.Offset, data);
            return 2;
        }
        else {
            tsCodeGen.addText(`[`);
            writeCodeWithQuotes(mapCode, sourceRange, data);
            tsCodeGen.addText(`]`);
            return 3;
        }
    }
    function writeCodeWithQuotes(mapCode, sourceRanges, data) {
        const addText = `'${mapCode}'`;
        for (const sourceRange of 'length' in sourceRanges ? sourceRanges : [sourceRanges]) {
            addMapping(tsCodeGen, {
                sourceRange,
                mappedRange: {
                    start: tsCodeGen.getText().length + 1,
                    end: tsCodeGen.getText().length + addText.length - 1,
                },
                mode: SourceMaps.Mode.Offset,
                additional: [
                    {
                        sourceRange,
                        mappedRange: {
                            start: tsCodeGen.getText().length,
                            end: tsCodeGen.getText().length + addText.length,
                        },
                        mode: SourceMaps.Mode.Totally,
                    }
                ],
                data,
            });
        }
        tsCodeGen.addText(addText);
    }
    function writeInterpolation(mapCode, sourceOffset, data, prefix, suffix) {
        (0, transform_1.walkInterpolationFragment)(ts, prefix + mapCode + suffix, (frag, fragOffset, isJustForErrorMapping) => {
            if (fragOffset === undefined) {
                tsCodeGen.addText(frag);
            }
            else {
                fragOffset -= prefix.length;
                let addSubfix = '';
                const overLength = fragOffset + frag.length - mapCode.length;
                if (overLength > 0) {
                    addSubfix = frag.substring(frag.length - overLength);
                    frag = frag.substring(0, frag.length - overLength);
                }
                if (fragOffset < 0) {
                    tsCodeGen.addText(frag.substring(0, -fragOffset));
                    frag = frag.substring(-fragOffset);
                    fragOffset = 0;
                }
                if (sourceOffset !== undefined && data !== undefined) {
                    writeCode(frag, {
                        start: sourceOffset + fragOffset,
                        end: sourceOffset + fragOffset + frag.length,
                    }, SourceMaps.Mode.Offset, isJustForErrorMapping
                        ? {
                            vueTag: data.vueTag,
                            capabilities: {
                                diagnostic: true,
                            },
                        }
                        : data);
                }
                else {
                    tsCodeGen.addText(frag);
                }
                tsCodeGen.addText(addSubfix);
            }
        }, localVars, identifiers);
    }
    function writeFormatCode(mapCode, sourceOffset, formatWrapper) {
        tsFormatCodeGen.addText(formatWrapper[0]);
        const targetRange = tsFormatCodeGen.addText(mapCode);
        addMapping(tsFormatCodeGen, {
            mappedRange: targetRange,
            sourceRange: {
                start: sourceOffset,
                end: sourceOffset + mapCode.length,
            },
            mode: SourceMaps.Mode.Offset,
            data: {
                vueTag: 'template',
                capabilities: {},
            },
        });
        tsFormatCodeGen.addText(formatWrapper[1]);
        tsFormatCodeGen.addText(`\n;\n`);
    }
    function writeCode(mapCode, sourceRange, mode, data) {
        const targetRange = tsCodeGen.addText(mapCode);
        addMapping(tsCodeGen, {
            sourceRange,
            mappedRange: targetRange,
            mode,
            data,
        });
    }
    function addMapping(gen, mapping) {
        var _a, _b;
        const newMapping = Object.assign({}, mapping);
        const templateStart = (_a = htmlToTemplate(mapping.sourceRange.start, mapping.sourceRange.end)) === null || _a === void 0 ? void 0 : _a.start;
        if (templateStart === undefined)
            return; // not found
        const offset = templateStart - mapping.sourceRange.start;
        newMapping.sourceRange = {
            start: mapping.sourceRange.start + offset,
            end: mapping.sourceRange.end + offset,
        };
        if (mapping.additional) {
            newMapping.additional = [];
            for (const other of mapping.additional) {
                let otherTemplateStart = (_b = htmlToTemplate(other.sourceRange.start, other.sourceRange.end)) === null || _b === void 0 ? void 0 : _b.start;
                if (otherTemplateStart === undefined)
                    continue;
                const otherOffset = otherTemplateStart - other.sourceRange.start;
                newMapping.additional.push(Object.assign(Object.assign({}, other), { sourceRange: {
                        start: other.sourceRange.start + otherOffset,
                        end: other.sourceRange.end + otherOffset,
                    } }));
            }
        }
        gen.addMapping2(newMapping);
    }
}
exports.generate = generate;
;
function toUnicode(str) {
    return str.split('').map(value => {
        var temp = value.charCodeAt(0).toString(16).padStart(4, '0');
        if (temp.length > 2) {
            return '\\u' + temp;
        }
        return value;
    }).join('');
}
function unHyphenatComponentName(newName) {
    return (0, shared_1.camelize)('-' + newName);
}
function keepHyphenateName(oldName, newName) {
    if (oldName === (0, shared_1.hyphenate)(oldName)) {
        return (0, shared_1.hyphenate)(newName);
    }
    return newName;
}
// https://github.com/vuejs/vue-next/blob/master/packages/compiler-dom/src/transforms/vModel.ts#L49-L51
// https://v3.vuejs.org/guide/forms.html#basic-usage
function getModelValuePropName(node, isVue2) {
    var _a;
    const tag = node.tag;
    const typeAttr = node.props.find(prop => prop.type === 6 /* ATTRIBUTE */ && prop.name === 'type');
    const type = (_a = typeAttr === null || typeAttr === void 0 ? void 0 : typeAttr.value) === null || _a === void 0 ? void 0 : _a.content;
    if (tag === 'input' && type === 'checkbox')
        return 'checked';
    if (tag === 'input' && type === 'radio')
        return '__VLS_radioBinding';
    if (tag === 'input' ||
        tag === 'textarea' ||
        tag === 'select' ||
        isVue2)
        return 'value';
    return 'modelValue';
}
// TODO: track https://github.com/vuejs/vue-next/issues/3498
function getPatchForSlotNode(node) {
    const forDirective = node.props.find((prop) => prop.type === 7 /* DIRECTIVE */
        && prop.name === 'for');
    if (forDirective) {
        let forNode;
        CompilerCore.processFor(node, forDirective, exports.transformContext, _forNode => {
            forNode = Object.assign({}, _forNode);
            return undefined;
        });
        if (forNode) {
            forNode.children = [Object.assign(Object.assign({}, node), { props: node.props.filter(prop => prop !== forDirective) })];
            return forNode;
        }
    }
}
exports.getPatchForSlotNode = getPatchForSlotNode;
function notEmpty(value) {
    return value !== null && value !== undefined;
}
