// @volar/vue-typescript/out/use/useSfcTemplateScript.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSfcTemplateScript = void 0;
const code_gen_1 = require("@volar/code-gen");
const templateGen = require("@volar/vue-code-gen/out/generators/template");
const reactivity_1 = require("@vue/reactivity");
const sourceMaps_1 = require("../utils/sourceMaps");
const string_1 = require("../utils/string");
const SourceMaps = require("@volar/source-map");
const path = require("path");
const transform_1 = require("@volar/vue-code-gen/out/transform");
const localTypes_1 = require("../utils/localTypes");
function useSfcTemplateScript(ts, fileName, template, script, scriptSetup, scriptSetupRanges, styles, styleFiles, styleEmbeddeds, templateData, sfcTemplateCompileResult, sfcStyles, scriptLang, compilerOptions, baseCssModuleType, getCssVBindRanges, getCssClasses, isVue2, disableTemplateScript) {
    const baseFileName = path.basename(fileName);
    const cssModuleClasses = (0, reactivity_1.computed)(() => styleFiles.value.reduce((obj, style) => {
        if (style.data.module) {
            const classes = getCssClasses(style);
            obj[style.data.module] = { [style.fileName]: classes };
        }
        return obj;
    }, {}));
    const cssScopedClasses = (0, reactivity_1.computed)(() => {
        var _a;
        const obj = {};
        const setting = (_a = compilerOptions.experimentalResolveStyleCssClasses) !== null && _a !== void 0 ? _a : 'scoped';
        for (const style of styleFiles.value) {
            if ((setting === 'scoped' && style.data.scoped) || setting === 'always') {
                const classes = getCssClasses(style);
                obj[style.fileName] = classes;
            }
        }
        return obj;
    });
    const templateCodeGens = (0, reactivity_1.computed)(() => {
        var _a, _b;
        if (!templateData.value)
            return;
        if (!((_a = sfcTemplateCompileResult.value) === null || _a === void 0 ? void 0 : _a.ast))
            return;
        return templateGen.generate(ts, templateData.value.lang, sfcTemplateCompileResult.value.ast, compilerOptions.experimentalCompatMode === 2, (_b = compilerOptions.experimentalRuntimeMode) !== null && _b !== void 0 ? _b : 'runtime-dom', !!compilerOptions.experimentalAllowTypeNarrowingInInlineHandlers, Object.values(cssScopedClasses.value).map(map => Object.keys(map)).flat(), templateData.value.htmlToTemplate, {
            getEmitCompletion: string_1.SearchTexts.EmitCompletion,
            getPropsCompletion: string_1.SearchTexts.PropsCompletion,
        });
    });
    const data = (0, reactivity_1.computed)(() => {
        const codeGen = new code_gen_1.CodeGen();
        codeGen.addText(`import * as __VLS_types from './__VLS_types';\n`);
        if (script.value || scriptSetup.value) {
            codeGen.addText(`import { __VLS_options, __VLS_name } from './${baseFileName}.__VLS_script';\n`);
            codeGen.addText(`import __VLS_component from './${baseFileName}.__VLS_script';\n`);
        }
        else {
            codeGen.addText(`var __VLS_name = undefined;\n`);
            codeGen.addText(`var __VLS_options = {};\n`);
            codeGen.addText(`var __VLS_component = (await import('${(0, localTypes_1.getVueLibraryName)(isVue2)}')).defineComponent({});\n`);
        }
        writeImportTypes();
        codeGen.addText(`declare var __VLS_ctx: InstanceType<typeof __VLS_component> & {\n`);
        /* CSS Module */
        const cssModuleMappingsArr = [];
        for (const moduleName in cssModuleClasses.value) {
            const moduleClasses = cssModuleClasses.value[moduleName];
            codeGen.addText(`${moduleName}: ${baseCssModuleType} & {\n`);
            cssModuleMappingsArr.push(writeCssClassProperties(moduleClasses, true, 'string', false));
            codeGen.addText('};\n');
        }
        codeGen.addText(`};\n`);
        codeGen.addText(`declare var __VLS_vmUnwrap: typeof __VLS_options & { components: { } };\n`);
        /* Components */
        codeGen.addText('/* Components */\n');
        codeGen.addText('declare var __VLS_otherComponents: NonNullable<typeof __VLS_component extends { components: infer C } ? C : {}> & __VLS_types.GlobalComponents & typeof __VLS_vmUnwrap.components & __VLS_types.PickComponents<typeof __VLS_ctx>;\n');
        codeGen.addText('declare var __VLS_ownComponent: __VLS_types.SelfComponent<typeof __VLS_name, typeof __VLS_component>;\n');
        codeGen.addText('declare var __VLS_allComponents: typeof __VLS_otherComponents & Omit<typeof __VLS_ownComponent, keyof typeof __VLS_otherComponents>;\n');
        codeGen.addText('declare var __VLS_rawComponents: __VLS_types.ConvertInvalidComponents<typeof __VLS_allComponents> & JSX.IntrinsicElements;\n'); // sort by priority
        codeGen.addText(`__VLS_allComponents.${string_1.SearchTexts.Components};\n`);
        codeGen.addText(`({} as __VLS_types.GlobalAttrs).${string_1.SearchTexts.GlobalAttrs};\n`);
        /* Style Scoped */
        codeGen.addText('/* Style Scoped */\n');
        codeGen.addText('type __VLS_StyleScopedClasses = {\n');
        const cssScopedMappings = writeCssClassProperties(cssScopedClasses.value, true, 'boolean', true);
        codeGen.addText('};\n');
        codeGen.addText('declare var __VLS_styleScopedClasses: __VLS_StyleScopedClasses | keyof __VLS_StyleScopedClasses | (keyof __VLS_StyleScopedClasses)[];\n');
        codeGen.addText(`/* CSS variable injection */\n`);
        writeCssVars();
        if (templateCodeGens.value) {
            (0, code_gen_1.mergeCodeGen)(codeGen, templateCodeGens.value.codeGen);
        }
        codeGen.addText(`export default __VLS_slots;\n`);
        return {
            codeGen,
            cssModuleMappingsArr,
            cssScopedMappings,
        };
        function writeImportTypes() {
            const bindingsArr = [];
            if (scriptSetupRanges.value && scriptSetup.value) {
                bindingsArr.push({
                    typeBindings: scriptSetupRanges.value.typeBindings,
                    content: scriptSetup.value.content,
                });
            }
            // if (scriptRanges.value && script.value) {
            // 	bindingsArr.push({
            // 		typeBindings: scriptRanges.value.typeBindings,
            // 		content: script.value.content,
            // 	});
            // }
            codeGen.addText('import {\n');
            for (const bindings of bindingsArr) {
                for (const typeBinding of bindings.typeBindings) {
                    const text = bindings.content.substring(typeBinding.start, typeBinding.end);
                    codeGen.addText(`__VLS_types_${text} as ${text},\n`);
                }
            }
            codeGen.addText(`} from './${baseFileName}.__VLS_script';\n`);
        }
        function writeCssClassProperties(data, patchRename, propertyType, optional) {
            const mappings = new Map();
            for (const uri in data) {
                const classes = data[uri];
                if (!mappings.has(uri)) {
                    mappings.set(uri, []);
                }
                for (const className in classes) {
                    const ranges = classes[className];
                    mappings.get(uri).push({
                        tsRange: {
                            start: codeGen.getText().length + 1,
                            end: codeGen.getText().length + 1 + className.length,
                        },
                        cssRanges: ranges,
                        mode: SourceMaps.Mode.Offset,
                        patchRename,
                    });
                    mappings.get(uri).push({
                        tsRange: {
                            start: codeGen.getText().length,
                            end: codeGen.getText().length + className.length + 2,
                        },
                        cssRanges: ranges,
                        mode: SourceMaps.Mode.Totally,
                        patchRename,
                    });
                    codeGen.addText(`'${className}'${optional ? '?' : ''}: ${propertyType},\n`);
                }
            }
            return mappings;
        }
        function writeCssVars() {
            const emptyLocalVars = {};
            const identifiers = new Set();
            for (let i = 0; i < sfcStyles.value.length; i++) {
                const style = sfcStyles.value[i];
                const binds = getCssVBindRanges(style);
                for (const cssBind of binds) {
                    const bindText = style.content.substring(cssBind.start, cssBind.end);
                    (0, transform_1.walkInterpolationFragment)(ts, bindText, (frag, fragOffset, isJustForErrorMapping) => {
                        if (fragOffset === undefined) {
                            codeGen.addText(frag);
                        }
                        else {
                            codeGen.addCode(frag, {
                                start: cssBind.start + fragOffset,
                                end: cssBind.start + fragOffset + frag.length,
                            }, SourceMaps.Mode.Offset, {
                                vueTag: 'style',
                                vueTagIndex: i,
                                capabilities: isJustForErrorMapping ? {
                                    diagnostic: true,
                                } : {
                                    basic: true,
                                    references: true,
                                    definitions: true,
                                    diagnostic: true,
                                    rename: true,
                                    completion: true,
                                    semanticTokens: true,
                                },
                            });
                        }
                    }, emptyLocalVars, identifiers);
                    codeGen.addText(';\n');
                }
            }
        }
    });
    const embedded = (0, reactivity_1.computed)(() => {
        var _a, _b;
        if (!disableTemplateScript && file.value) {
            const sourceMap = new SourceMaps.SourceMapBase(data.value.codeGen.getMappings(parseMappingSourceRange));
            for (const [fileName, mappings] of [
                ...data.value.cssModuleMappingsArr.flatMap(m => [...m]),
                ...data.value.cssScopedMappings,
            ]) {
                const cssSourceMap = (_a = styleEmbeddeds.value.find(embedded => embedded.file.fileName === fileName)) === null || _a === void 0 ? void 0 : _a.sourceMap;
                if (!cssSourceMap)
                    continue;
                for (const mapped of mappings) {
                    const tsRange = mapped.tsRange;
                    for (const cssRange of mapped.cssRanges) {
                        const vueRange = (_b = cssSourceMap.getSourceRange(cssRange.start, cssRange.end)) === null || _b === void 0 ? void 0 : _b[0];
                        if (!vueRange)
                            continue;
                        sourceMap.mappings.push({
                            data: {
                                vueTag: 'style',
                                capabilities: {
                                    references: true,
                                    rename: true,
                                    referencesCodeLens: mapped.mode === SourceMaps.Mode.Totally, // has 2 modes
                                },
                                normalizeNewName: mapped.patchRename ? beforeCssRename : undefined,
                                applyNewName: mapped.patchRename ? doCssRename : undefined,
                            },
                            mode: mapped.mode,
                            sourceRange: vueRange,
                            mappedRange: tsRange,
                        });
                    }
                }
            }
            return {
                file: file.value,
                sourceMap,
            };
        }
    });
    const formatEmbedded = (0, reactivity_1.computed)(() => {
        if (templateCodeGens.value && formatFile.value) {
            const sourceMap = new sourceMaps_1.EmbeddedFileSourceMap(templateCodeGens.value.formatCodeGen.getMappings(parseMappingSourceRange));
            return {
                file: formatFile.value,
                sourceMap,
            };
        }
    });
    const inlineCssFile = (0, reactivity_1.computed)(() => {
        if (templateCodeGens.value) {
            const file = {
                fileName: fileName + '.template.css',
                lang: 'css',
                content: templateCodeGens.value.cssCodeGen.getText(),
                capabilities: {
                    diagnostics: false,
                    foldingRanges: false,
                    formatting: false,
                    codeActions: false,
                    documentSymbol: false,
                    inlayHints: false,
                },
                data: undefined,
                // data: {
                // 	module: false,
                // 	scoped: false,
                // },
                isTsHostFile: false,
            };
            return file;
        }
    });
    const inlineCssEmbedded = (0, reactivity_1.computed)(() => {
        if (templateCodeGens.value && inlineCssFile.value) {
            const sourceMap = new sourceMaps_1.EmbeddedFileSourceMap(templateCodeGens.value.cssCodeGen.getMappings(parseMappingSourceRange));
            return {
                file: inlineCssFile.value,
                sourceMap,
            };
        }
    });
    const file = (0, reactivity_1.computed)(() => {
        if (data.value) {
            const lang = scriptLang.value === 'js' ? 'jsx' : scriptLang.value === 'ts' ? 'tsx' : scriptLang.value;
            const embeddedFile = {
                fileName: fileName + '.__VLS_template.' + lang,
                lang: lang,
                content: data.value.codeGen.getText(),
                capabilities: {
                    diagnostics: true,
                    foldingRanges: false,
                    formatting: false,
                    documentSymbol: false,
                    codeActions: false,
                    inlayHints: true,
                },
                data: undefined,
                isTsHostFile: true,
            };
            return embeddedFile;
        }
    });
    const formatFile = (0, reactivity_1.computed)(() => {
        if (templateCodeGens.value) {
            const lang = scriptLang.value === 'js' ? 'jsx' : scriptLang.value === 'ts' ? 'tsx' : scriptLang.value;
            const embeddedFile = {
                fileName: fileName + '.__VLS_template.format.' + lang,
                lang: lang,
                content: templateCodeGens.value.formatCodeGen.getText(),
                capabilities: {
                    diagnostics: false,
                    foldingRanges: false,
                    formatting: true,
                    documentSymbol: true,
                    codeActions: false,
                    inlayHints: false,
                },
                data: undefined,
                isTsHostFile: false,
            };
            return embeddedFile;
        }
    });
    return {
        templateCodeGens,
        embedded,
        file,
        formatFile,
        formatEmbedded,
        inlineCssFile,
        inlineCssEmbedded,
    };
    function parseMappingSourceRange(data, range) {
        var _a, _b;
        if ((data === null || data === void 0 ? void 0 : data.vueTag) === 'style' && (data === null || data === void 0 ? void 0 : data.vueTagIndex) !== undefined) {
            return {
                start: styles.value[data.vueTagIndex].startTagEnd + range.start,
                end: styles.value[data.vueTagIndex].startTagEnd + range.end,
            };
        }
        const templateOffset = (_b = (_a = template.value) === null || _a === void 0 ? void 0 : _a.startTagEnd) !== null && _b !== void 0 ? _b : 0;
        return {
            start: templateOffset + range.start,
            end: templateOffset + range.end,
        };
    }
}
exports.useSfcTemplateScript = useSfcTemplateScript;
function beforeCssRename(newName) {
    return newName.startsWith('.') ? newName.slice(1) : newName;
}
function doCssRename(oldName, newName) {
    return '.' + newName;
}
