// @volar/vue-code-gen/out/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.compileSFCTemplate = exports.generateSFCScriptTypeCheckCode = void 0;
const script_1 = require("./generators/script");
const scriptRanges_1 = require("./parsers/scriptRanges");
const scriptSetupRanges_1 = require("./parsers/scriptSetupRanges");
const CompilerDOM = require("@vue/compiler-dom");
const CompilerVue2 = require("./vue2TemplateCompiler");
__exportStar(require("./types"), exports);
__exportStar(require("@vue/compiler-dom"), exports);
/**
 * @param templateAst Use `require('@vue/compiler-dom').compile` or `require('@volar/vue-code-gen').compileTemplate`, provide to resolve variables unused in script setup
 * @param cssVars Use `require('@vue/compiler-sfc').parseCssVars`, provide to resolve variables unused in script setup
 * @param vueLibName Where should `defineComponent` and `PropType` import from? (For example: `vue`, `@vue/runtime-dom`, `@vue/composition-api`)
 */
function generateSFCScriptTypeCheckCode(ts, scriptLang, scriptCode, scriptSetupCode, shimComponentOptions, downgradePropsAndEmitsToSetupReturnOnScriptSetup, templateAst, cssVars, vueLibName = 'vue') {
    const generated = (0, script_1.generate)('script', '', scriptCode !== undefined ? { content: scriptCode } : undefined, scriptSetupCode !== undefined ? { content: scriptSetupCode } : undefined, scriptCode !== undefined ? (0, scriptRanges_1.parseScriptRanges)(ts, ts.createSourceFile('dummy.' + scriptLang, scriptCode, ts.ScriptTarget.ESNext), scriptSetupCode !== undefined, false, false) : undefined, scriptSetupCode !== undefined ? (0, scriptSetupRanges_1.parseScriptSetupRanges)(ts, ts.createSourceFile('dummy.' + scriptLang, scriptSetupCode, ts.ScriptTarget.ESNext)) : undefined, () => undefined, // TODO
    // () => templateAst ? generateTemplateScript(templateAst) : undefined,
    () => cssVars !== null && cssVars !== void 0 ? cssVars : [], vueLibName, shimComponentOptions, downgradePropsAndEmitsToSetupReturnOnScriptSetup);
    return {
        code: generated.codeGen.getText(),
        scriptMappings: getScriptMappings('script'),
        scriptSetupMappings: getScriptMappings('scriptSetup'),
    };
    function getScriptMappings(vueTag) {
        return generated.codeGen.getMappings()
            .filter(mapping => mapping.data.vueTag === vueTag
            && mapping.data.capabilities.diagnostic)
            .map(mapping => ({
            sourceTextRange: mapping.sourceRange,
            generatedTextRange: mapping.mappedRange,
        }));
    }
}
exports.generateSFCScriptTypeCheckCode = generateSFCScriptTypeCheckCode;
/**
 * A wrapper function of `require('@vue/compiler-dom').compile`
 */
function compileSFCTemplate(htmlCode, options = {}, vueVersion = 3) {
    const errors = [];
    const warnings = [];
    let ast;
    try {
        ast = (vueVersion === 2 ? CompilerVue2 : CompilerDOM).compile(htmlCode, Object.assign({ onError: (err) => errors.push(err), onWarn: (err) => warnings.push(err), expressionPlugins: ['typescript'] }, options)).ast;
    }
    catch (e) {
        const err = e;
        errors.push(err);
    }
    return {
        errors,
        warnings,
        ast,
    };
}
exports.compileSFCTemplate = compileSFCTemplate;
